docs/designs/ruby_test_plan_002.md

```diff
@@ -0,0 +1,116 @@
+support the most widespread and safe ones instead of supporting everything.
+
+The tests are to check that conforming packages could be preprocessed by Cachi2
+and that non-conforming packages are rejected by Cachi2.
+
+The checks are expected to be run against actual packages published either on
+rubygems.org, on github.com or locally.
+
+Each example in every test scenario requires a separate repository/package
+demonstrating an aspect under test. A list of necessary repositories could be
+found [below](#necessary-repositories).
+
+## Test cases
+
+@e2e_test
+Feature: Projects prepared with Cachi2 could be hermetically built
+    Scenario: a user can successfully build a project prepared with Cachi2
+        Given a clean repository of a well-defined project
+         When a user request to fetch dependencies for a well-defined project
+          And a user generates environment for well-defined project
+          And a user triggers a hermetic build
+         Then the build system exits successfully
+          And a build artifact is produced.
+
+
+@integration_test
+Feature: Cachi2 can prefetch dependencies for Ruby projects
+    
+    Scenario outline: a user wants to fetch dependencies for a package containing well-defined Gemfile.lock
+       Given  clean repository of a "well-defined project"
+        When  a user request to fetch dependencies for a "well-defined project"
+        Then  direct and transitive dependencies are fetched
+         And  dependencies could be found within the "well-defined project"
+    Examples:
+        |    well-defined project                                       |
+        | Has a mixture of gems, Git and path dependencies and .gemspec |
+        | Has a mixture of gems, Git and path dependencies              |
+
+    Scenario outline: a user wants to fetch dependencies for a package containing ill-defined Gemfile.lock
+       Given  clean repository of a "ill-defined project"
+        When  a user request to fetch dependencies for a "ill-defined project"
+        Then  Cachi2 reports an error
+         And  repository for "ill-defined project" remains unchanged
+    Examples:
+        |    ill-defined project                           |
+        | Gemfile.lock is missing                          |
+        | Git dependency is not pinned to a revision       |
+        | Transitive dependency is ill-defined             |
+
+
+@integration_test
+Feature: Cachi2 can preset configuration for hermetic builds
+
+    Scenario outline: a user prepares a repository with a Ruby project
+        When  a user generates environment for "a Ruby project"
+        Then  the environment contains a pointer to local Bundler config
+         And  Bundler cache directory structure is recreated for every project
+    Examples:
+        |    a Ruby project                                       |
+        | Single-project repository with mixed depenendcies types |
```

Comment:
I wonder if we could combine the single-project repo case + bundler configuration with the previous positive test case since based on the configuration opts order of precedence we have to dump a config file either way, the only question is whether it'll be combined with user options or just ours, so I'd create the previous positive case such that user defines some options in a custom config and cover  only the more complex scenario where a user config option might potentially collide with ours. Note that in order to test this we'll likely have to use the `inject-files` command which also means we need to run a prefetch first which feels redundant to me when the previous scenario has to already do that, so I'd prefer if we ran as few test cases as possible and combine whatever is related together (in general).

Reply:
I don't think combining multiple test scenario together is a good idea. Each defined scenario should test one distinct aspect. The need to clone a repo or the need to run as few tests as possible is an implementation detail and combining multiple tests together is essentially leaking implementation details into higher-level. From what I have seen on the gate the tests tend to run for just few seconds, or tens of seconds when podman has to build something. I'd say optimizations should be handled by tests implementation: all expensive operations could be lumped together into a fixture or any other form of shared context. On the implementation level things could and likely should be combined for efficiency.

To make sure that this is not forgotten by an implementer I'll add a couple of sentences about a fixture or joining scenarios to the end of section.

Reply:
Agreed. I spent days trying to understand/fix some pip unit tests because they were *testing everything*.

Eventually I was forced to simply refactor them out based on type under test, which actually tested the same stuff in fewer SLOC because conditionals based on behavior/data differences in type were no longer needed.

Reply:
> Agreed. I spent days trying to understand/fix some pip unit tests because they were testing everything.

Wasn't this due to the matrix of parametrize? That's different though, because you have a matrix that explodes in complexity.

> I don't think combining multiple test scenario together is a good idea. Each defined scenario should test one distinct aspect.

I don't dispute this, but what we're really talking about here is to add essentially one more assert after running `inject-files` which is IMO still a constant complexity since we're not introducing any matrix for the prefetch test to handle. I wouldn't suggest it if we didn't have the data for a successful `inject-files` check already at hand.

It's not a blocker for me though if we keep it a separate scenario, I'm just always wary of the way we structure integration tests simply because of how much time they take to execute, e.g. the e2e's are painfully slow and based on the fact that this is inherently only going to get worse we can only control or tame the pace of it.

Reply:
I think we are talking about pretty much the same thing: high-level test cases should remain cleanly separated, but at the implementation level could and should be optimized. I added a note at L117 in the latest revision to stress that.

Reply:
> Wasn't this due to the matrix of parametrize

That was one aspect, yes. The sheer amount of stuff being tested was the main issue.